### PR TITLE
Add an in-out boolean color-scheme property to Palette

### PR DESCRIPTION
**Pull request by tronical**
This allows applications to force dark/light mode, as well as determine which mode is active.

### CODE REVIEW

**Comment by tronical** [comment]
Another option would be to have an enum instead of a boolean, like `enum ColorScheme { Dark, Light, Unknown }`, but that's a little weird for the setter.

File: internal/backends/qt/qt_widgets/palette.rs [hunk 1]
```diff
@@ -41,6 +41,7 @@ pub struct NativePalette {
     pub selection_background: Property<Brush>,
     pub selection_foreground: Property<Brush>,
     pub border: Property<Brush>,
+    pub dark_color_scheme: Property<bool>, // This is a no-op as Qt doesn't allow changing the palette yet.
```
Comment (ogoffart) [inline internal/backends/qt/qt_widgets/palette.rs:44]: This should then use the `StyleMetrics.dark-color-scheme`  which is initialized depending on the Palette's color

File: docs/reference/src/language/builtins/globals.md [hunk 2]
```diff
@@ -19,6 +19,8 @@ the selected style e.g. fluent, cupertino, material, or qt.
 -   **`selection-background`** (_out_ _brush_): Defines the background brush that is used to highlight a selection such as a text selection.
 -   **`selection-foreground`** (_out_ _brush_):  Defines the foreground brush that is used for content that is displayed on `selection-background` brush.
 -   **`border`** (_out_ _brush_): Defines the brush that is used for borders such as separators and widget borders.
+-   **`dark-color-scheme`** (_in_ _out_ _bool_): Read this property to determine if the currently selected palette follows a dark color scheme.
+    Set this property to force a dark or light color scheme. All styles except for the Qt style support setting a dark or light color scheme.
```
Comment (ogoffart) [inline docs/reference/src/language/builtins/globals.md:23]: do the -dark style support setting a light color style and vice versa?

**Review by ogoffart — COMMENTED** [review]
Can we have a switch in the gallery?

File: docs/reference/src/language/builtins/globals.md (hunk 2 above)
Comment (tronical) [reply docs/reference/src/language/builtins/globals.md:23]: Fortunately yes. Basically they work like this:

```
global FluentPalette {
    in-out property <bool> dark-color-scheme: ColorSchemeSelector.dark-color-scheme;
```

and `ColorSchemeSelector`'s property is either true or false depending on the chosen style (-light/-dark suffix) or it defaults to `SlintInternal.dark-color-scheme` (the builtin that calls the window adapter). But everything in the fluent style uses the above property, so the exposed `Palette.dark-color-scheme`'s two-way binding ends up breaking up the binding to `ColorSchemeSelector` and setting the forced value (after which it doesn't react to system changes anymore).

File: internal/backends/qt/qt_widgets/palette.rs (hunk 1 above)
Comment (tronical) [reply internal/backends/qt/qt_widgets/palette.rs:44]: Ah yes, good point. Too bad I can't access the instances :(. I guess I have to duplicate the code.

**Comment by tronical** [comment]
> Can we have a switch in the gallery?

Good idea - done :)

**Comment by FloVanGH** [comment]
> Another option would be to have an enum instead of a boolean, like `enum ColorScheme { Dark, Light, Unknown }`, but that's a little weird for the setter.

That was already my suggestion. I think this makes it much more flexible and easier to extend it in the future. There are often now different color schemes e.g. for better accessibility e.g. HighContrast or a scheme for people with green-blue deficiency.

**Comment by FloVanGH** [comment]
> > Can we have a switch in the gallery?
> 
> Good idea - done :)

Yeah it's a good idea. It enhances the interactivity of the gallery. I had done the same on my initial PR https://github.com/slint-ui/slint/blob/357eb48cbd2eaa198aaa48398fd1b7fb5841fa12/examples/gallery/ui/pages/page.slint#L31

**Comment by tronical** [comment]
> > Another option would be to have an enum instead of a boolean, like `enum ColorScheme { Dark, Light, Unknown }`, but that's a little weird for the setter.
> 
> That was already my suggestion. I think this makes it much more flexible and easier to extend it in the future. There are often now different color schemes e.g. for better accessibility e.g. HighContrast or a scheme for people with green-blue deficiency.

Right, I've noticed the same. The main issue I have with `Unknown` is ... what should happen when it is set? :)

**Comment by FloVanGH** [comment]
> > > Another option would be to have an enum instead of a boolean, like `enum ColorScheme { Dark, Light, Unknown }`, but that's a little weird for the setter.
> > 
> > 
> > That was already my suggestion. I think this makes it much more flexible and easier to extend it in the future. There are often now different color schemes e.g. for better accessibility e.g. HighContrast or a scheme for people with green-blue deficiency.
> 
> Right, I've noticed the same. The main issue I have with `Unknown` is ... what should happen when it is set? :)

Do we need `Unknown` I think if the system does not define a scheme a default fallback should be used e.g. `Light`.

**Comment by tronical** [comment]
Fair point :-)

**Comment by ogoffart** [comment]
> The main issue I have with Unknown is ... what should happen when it is set? :)

I'd say it should then be back to the global desktop or browser setting.

File: internal/compiler/widgets/cosmic-base/std-widgets-impl.slint [hunk 3]
```diff
@@ -24,7 +24,6 @@ export global StyleMetrics  {
     out property <color> textedit-text-color: CosmicPalette.foreground;
     out property <color> textedit-background-disabled: CosmicPalette.control-disabled;
     out property <color> textedit-text-color-disabled: CosmicPalette.text-disabled;
-    out property <bool> dark-color-scheme: CosmicPalette.dark-color-scheme;
```
Comment (tronical) [inline internal/compiler/widgets/cosmic-base/std-widgets-impl.slint:27]: This is - strictly speaking - a breaking change. StyleMetrics is undocumented, but accessible to apps. I'd rather keep things thing, but if you all feel like we should keep it, then I can keep this one here for compatibility.

**Comment by tronical** [comment]
Alright, so there's an enum now. Initially `Palette.color-scheme` is bound to `SlintInternal.color-scheme`, so it'll return whatever we detect in the system. Setting `Palette.color-scheme` to `ColorScheme.unknown` now makes the styles fall back to `SlintInternal.color-scheme`, i.e. the system default, but when reading the property it'll return "unknown".

File: internal/compiler/widgets/cosmic-base/std-widgets-impl.slint (hunk 3 above)
Comment (ogoffart) [reply internal/compiler/widgets/cosmic-base/std-widgets-impl.slint:27]: There is a lot of users out there:
https://github.com/search?q=StyleMetrics.dark-color-scheme+-repo%3Aslint-ui%2Fslint&type=code
Please keep it.

If anything we could deprecate it somehow.

Comment (tronical) [reply internal/compiler/widgets/cosmic-base/std-widgets-impl.slint:27]: I've restored it and edited the description of https://github.com/slint-ui/slint/issues/3023

**Review by ogoffart — APPROVED** [review]

**Pull request merged**